### DISCUSSION TITLE
Options for handling contextual symbol types

### DISCUSSION THREAD

**Discussion by Cidolfas** [tree-sitter/tree-sitter#1806] [Q&A] [answered]
I'm writing a narrative game language, and the need to reduce friction with adding plenty of text spans is making a few bits of grammar tricky for me.

The core of the issue is that, depending on context, a span of symbols may want to be specific types (e.g. int, identifier) while other times we want to interpret them as a string.

Here's an example of the issue in my language:
```
@quality statue_facing
  #name The Statue's Gaze
  #type enum
  #value 0 entrance|name:The Entrance
  #value 1 pool|name:The Brackish Pool
```
This is defining an object that's a game value with the logical name 'statue_facing', whose display name is 'The Statue's Gaze', and which can hold one of two values. The value definitions require a specific syntax: an int, followed by a valid logical name. The key/value pairs after the bar are optional, but define a display name for those values.

Right now I'm parsing all properties - lines which start with a # symbol - in generic. Here's a stripped-down version of the grammar:
``` JavaScript
property: $ => seq(
  '#',
  $.name,
  $.token_span,
  optional(repeat(seq('|', $.modifier_pair)))
)

token_span: $ => repeat1(choice(
  $.int,
  $.name,
  prec(-1, /[^\|\r\n \t]+/)
))

modifier_pair: $ => seq(
  $.name,
  ':',
  $.token_span
)
```
What I'd like to do is, after the parser determines that we've got a property node, toss it into a function that would let me transform the token_span into a different set of elements - e.g. if the property's name is 'value' turn it into an 'int' node and a 'name' node, but if the property's name is 'name' then turn it into a single text_span node. But tree-sitter doesn't work this way.

I've been dealing with these rules so far at a level above the parser, by taking tree-sitter's parse tree and turning it into a different AST in my own code. But this loses some of the nice tools that tree-sitter provides - like syntax highlighting and being able to just reparse changed sections. (I can get the scoped reparsing on the underlying parse tree, but not my AST.)

I know that I could replicate this by defining each possible property type in specific, so instead of the property rule I'd have a property_name rule and a property_value rule whose grammars did have that level of specificity. But the plan is for users to be able to extend the known types of properties beyond the fixed base set.

I've also considered having the user-added known types be included into a generated syntax file generation, then generate the compiler from that auto-generated syntax. But that seems like a complicated pipeline.

Do I have other options here?

**Comment by ahelwer** [comment]
Use an external scanner to define whatever (fuzzy?) logic you want. If you can't write it that way then it can't be written.

**Reply by Cidolfas** [reply]
External scanners work by consuming character streams and emitting a single symbol, right? So I couldn't lean on any of my existing rules inside the external scanner, and it doesn't allow me to generate child nodes for that symbol?

**Reply by ahelwer** [reply, marked as answer]
I will often use external scanners to emit unique invisible zero-width tokens like `$._start_x_section` or `$._end_x_section` which are consumed by the grammar to shunt it into a certain rule bracketed by those tokens.

**Reply by Cidolfas** [reply]
The zero-width token idea doesn't seem to help me with this problem, but it did help me solve another issue in my grammar.

**Reply by Cidolfas** [reply]
I can't actually seem to create a zero-width token using an external scanner.

It works as expected when it is proceeded by whitespace I advance through, but if it's at the start of the file it'll chomp the first character always. Is there some trick to it?

**Reply by ahelwer** [reply]
If it only happens at the very start of the file the external scanner might be viewing the character it consumes as the byte order mark. Possibly a bug you can file.
